src/webgpu/api/operation/memory_sync/buffer.spec.ts

```diff
@@ -0,0 +1,261 @@
+      fn frag_main() -> void {
+	data.a = ${value};
+	outColor = vec4<f32>(1.0, 0.0, 0.0, 1.0);
+        return;
+      }
+
+      entry_point fragment = frag_main;
+    `;
+
+    return this.device.createRenderPipeline({
+      vertexStage: {
+        module: this.device.createShaderModule({
+          code: wgslVertex,
+        }),
+        entryPoint: 'vert_main',
+      },
+      fragmentStage: {
+        module: this.device.createShaderModule({
+          code: wgslFragment,
+        }),
+        entryPoint: 'frag_main',
+      },
+      primitiveTopology: 'point-list',
+      colorStates: [{ format: 'rgba8unorm' }],
+    });
+  }
+
+  beginSimpleRenderPass(encoder: GPUCommandEncoder): GPURenderPassEncoder {
+    const view = this.device
+      .createTexture({
+        size: { width: 4, height: 4, depth: 1 },
+        format: 'rgba8unorm',
+        usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+      })
+      .createView();
+    return encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: view,
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+        },
+      ],
+    });
+  }
+
+  // Write storage buffer via draw call in render pass. Use bundle if needed.
+  writeByRenderPass(bundle: boolean, buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteRenderPipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+    const pass = this.beginSimpleRenderPass(encoder);
+    if (bundle) {
+      const bundleEncoder = this.device.createRenderBundleEncoder({
+        colorFormats: ['rgba8unorm'],
+      });
+      bundleEncoder.setBindGroup(0, bindGroup);
+      bundleEncoder.setPipeline(pipeline);
+      bundleEncoder.draw(1, 1, 0, 0);
+      const bundles = bundleEncoder.finish();
+      pass.executeBundles([bundles]);
+    } else {
+      pass.setPipeline(pipeline);
+      pass.setBindGroup(0, bindGroup);
+      pass.draw(1, 1, 0, 0);
+    }
+    pass.endPass();
+  }
+
+  // Write storage buffer via dispatch call in compute pass.
+  writeByComputePass(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteComputePipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bindGroup);
+    pass.dispatch(1);
+    pass.endPass();
+  }
+
+  // Issue write operation via render pass, compute pass, copy, etc.
+  issueWriteOp(
+    writeOp: string,
+    bundle: boolean,
+    buffer: GPUBuffer,
+    value: number,
+    encoder: GPUCommandEncoder
+  ) {
+    const isRenderPass = writeOp === 'render';
+    assert(!bundle || isRenderPass);
+    if (isRenderPass) {
+      this.writeByRenderPass(bundle, buffer, value, encoder);
+    } else {
+      assert(writeOp === 'compute');
+      this.writeByComputePass(buffer, value, encoder);
+    }
```

Comment:
We can add a few more if-else to cover other wrietOps like t2b copy, b2b copy, writeBuffer, etc.

Reply:
How do you handle the fact that some of them happen on passes, some on the command encoder, and some at the device-level? I worry there are going to be a lot if if-else that bobble information up and down the stack to know whether to finish passes, command buffers etc.

Basically the current approach seems to work fine for two types of operations that are both on passes but extensibility to the full test seems like it will result in very complicated and branchy code.

Reply:
Maybe a helper class abstracting the encoder types and operations would work here?

Reply:
I use different levels of wrapper functions instead, it's c/c++-style though. But the logic is easy to understand too.